Improve chunkset_avx2 performance

KungFuJesus opened the pull request:
Make chunkset_avx2 half chunk aware
    
This gives us appreciable gains on a number of fronts.  The first being
we're inlining a pretty hot function that was getting dispatched to
regularly. Another is that we're able to do a safe lagged copy of a
distance that is smaller, so CHUNKCOPY gets its teeth back here for
smaller sizes, without having to do another dispatch to a function.
    
We're also now doing two overlapping writes at once and letting the CPU
do its store forwarding. This was an enhancement @dougallj had suggested
a while back.
   
Additionally, the "half chunk mag" here is fundamentally less
complicated because it doesn't require sythensizing cross lane permutes
with a blend operation, so we can optimistically do that first if the
len is small enough that a full 32 byte chunk doesn't make any sense.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced new inline functions for handling smaller data chunks using AVX2 instructions.
	- Added support for `halfchunk_t` type to optimize operations on 128-bit chunks.
	- Implemented a new function `HALFCHUNKCOPY` for efficient data copying of half-sized chunks.

- **Bug Fixes**
	- Adjusted memory operation handling in the `inflate_fast` function for improved safety.

- **Refactor**
	- Removed the `CHUNK_SIZE` macro across multiple files to streamline chunk size references.
	- Updated `CHUNKMEMSET` function to a static inline function for better linkage control.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

KungFuJesus committed:
Simplify avx2 chunkset a bit

Put length 16 in the length checking ladder and take care of it there
since it's also a simple case to handle. We kind of went out of our way
to pretend 128 bit vectors didn't exist when using avx2 but this can be
handled in a single instruction. Strangely the intrinsic uses vector
register operands but the instruction itself assumes a memory operand
for the source. This also means we don't have to handle this case in our
"GET_CHUNK_MAG" function.

KungFuJesus committed:
Make chunkset_avx2 half chunk aware

This gives us appreciable gains on a number of fronts.  The first being
we're inlining a pretty hot function that was getting dispatched to
regularly. Another is that we're able to do a safe lagged copy of a
distance that is smaller, so CHUNKCOPY gets its teeth back here for
smaller sizes, without having to do another dispatch to a function.

We're also now doing two overlapping writes at once and letting the CPU
do its store forwarding. This was an enhancement @dougallj had suggested
a while back.

Additionally, the "half chunk mag" here is fundamentally less
complicated because it doesn't require sythensizing cross lane permutes
with a blend operation, so we can optimistically do that first if the
len is small enough that a full 32 byte chunk doesn't make any sense.

Dead2 commented:
I tested current develop (with split chunkcopy_safe) with and without this PR on an i9-9900K.

Develop:
```
   text    data     bss     dec     hex filename
 135966    1312       8  137286   21846 libz-ng.so.2
          5,733.90 msec task-clock:u                     #    1.000 CPUs utilized               ( +-  0.01% )
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
               143      page-faults:u                    #   24.939 /sec                        ( +-  0.30% )
    19,908,943,906      cycles:u                         #    3.472 GHz                         ( +-  0.01% )
    35,633,452,503      instructions:u                   #    1.79  insn per cycle              ( +-  0.00% )
     3,749,361,374      branches:u                       #  653.894 M/sec                       ( +-  0.00% )
       181,427,372      branch-misses:u                  #    4.84% of all branches             ( +-  0.04% )

          5.734291 +- 0.000584 seconds time elapsed  ( +-  0.01% )
```

This PR:
```
 Performance counter stats for 'build/minigzip -c -d -k ../cov-analysis-linux64-2023.6.2.tar.gz' (10 runs):

          5,396.14 msec task-clock:u                     #    1.000 CPUs utilized               ( +-  0.02% )
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
               144      page-faults:u                    #   26.686 /sec                        ( +-  0.33% )
    18,752,218,475      cycles:u                         #    3.475 GHz                         ( +-  0.02% )
    35,633,297,945      instructions:u                   #    1.90  insn per cycle              ( +-  0.00% )
     3,748,055,372      branches:u                       #  694.580 M/sec                       ( +-  0.00% )
       175,467,241      branch-misses:u                  #    4.68% of all branches             ( +-  0.04% )

          5.396606 +- 0.000948 seconds time elapsed  ( +-  0.02% )
```
Pretty nice improvement to instructions/cycle.
Inflate is a whopping 5.8% faster on average.

Benchmark of this PR without #1776:  (Correct build used)
```
 Performance counter stats for 'build/minigzip -c -d -k ../cov-analysis-linux64-2023.6.2.tar.gz' (10 runs):

          5,888.93 msec task-clock:u                     #    1.000 CPUs utilized               ( +-  0.01% )
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
               143      page-faults:u                    #   24.283 /sec                        ( +-  0.21% )
    20,457,192,322      cycles:u                         #    3.474 GHz                         ( +-  0.01% )
    35,730,996,426      instructions:u                   #    1.75  insn per cycle              ( +-  0.00% )
     3,757,474,157      branches:u                       #  638.058 M/sec                       ( +-  0.00% )
       181,224,621      branch-misses:u                  #    4.82% of all branches             ( +-  0.01% )

          5.889362 +- 0.000682 seconds time elapsed  ( +-  0.01% )
```

---

Develop:
```
# x86-64 split_chunkcopy
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      1.203/1.221/1.229/0.008        0.415/0.432/0.438/0.006       94,127,485
 2     35.519%      2.253/2.272/2.283/0.009        0.427/0.435/0.439/0.004       75,286,310
 3     33.844%      2.830/2.847/2.857/0.010        0.404/0.410/0.415/0.003       71,735,206
 4     33.146%      3.147/3.185/3.203/0.016        0.391/0.399/0.405/0.004       70,255,211
 5     32.642%      3.508/3.551/3.583/0.024        0.383/0.394/0.402/0.006       69,187,407
 6     32.483%      4.025/4.105/4.132/0.025        0.384/0.394/0.398/0.004       68,850,764
 7     32.255%      5.981/6.024/6.058/0.023        0.387/0.394/0.397/0.003       68,366,747
 8     32.167%      8.673/8.730/8.764/0.028        0.387/0.392/0.396/0.003       68,180,750
 9     31.887%   12.039/12.099/12.165/0.036        0.377/0.389/0.395/0.005       67,586,430

 avg1  34.261%                        4.892                          0.404
 tot                                660.487                         54.599      653,576,310

   text    data     bss     dec     hex filename
 135966    1312       8  137286   21846 libz-ng.so.2
 ```
 
This PR:
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      1.203/1.219/1.231/0.008        0.421/0.429/0.433/0.003       94,127,485
 2     35.519%      2.261/2.281/2.289/0.007        0.420/0.433/0.439/0.005       75,286,310
 3     33.844%      2.828/2.850/2.865/0.012        0.402/0.407/0.412/0.003       71,735,206
 4     33.146%      3.138/3.189/3.217/0.025        0.394/0.403/0.408/0.004       70,255,211
 5     32.642%      3.493/3.549/3.578/0.025        0.384/0.393/0.398/0.005       69,187,407
 6     32.483%      4.051/4.109/4.140/0.027        0.388/0.395/0.400/0.004       68,850,764
 7     32.255%      5.957/6.024/6.059/0.030        0.383/0.392/0.397/0.004       68,366,747
 8     32.167%      8.717/8.740/8.761/0.013        0.380/0.390/0.394/0.004       68,180,750
 9     31.887%   12.054/12.108/12.153/0.030        0.383/0.388/0.391/0.003       67,586,430

 avg1  34.261%                        4.897                          0.403
 tot                                661.036                         54.447      653,576,310

   text    data     bss     dec     hex filename
 136350    1312       8  137670   219c6 libz-ng.so.2
```
For some reason the improvements were much smaller here, only 0.28% faster inflate on average.

The difference is speeds seems to be data-dependent, not cacheline-aliasing or anything like that.
(I verified this using different builds to ensure things get compiled and placed slightly differently, as I often do when in doubt)

---

Benchmark of this PR without #1776:  (Correct build used)
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      1.209/1.223/1.231/0.007        0.442/0.459/0.465/0.007       94,127,485
 2     35.519%      2.262/2.279/2.289/0.008        0.447/0.460/0.464/0.005       75,286,310
 3     33.844%      2.838/2.862/2.876/0.011        0.428/0.436/0.440/0.003       71,735,206
 4     33.146%      3.156/3.206/3.235/0.023        0.412/0.422/0.427/0.004       70,255,211
 5     32.642%      3.523/3.560/3.581/0.018        0.408/0.418/0.423/0.004       69,187,407
 6     32.483%      4.071/4.119/4.145/0.023        0.404/0.416/0.422/0.005       68,850,764
 7     32.255%      5.951/6.028/6.085/0.045        0.405/0.414/0.420/0.005       68,366,747
 8     32.167%      8.692/8.747/8.779/0.031        0.404/0.414/0.419/0.005       68,180,750
 9     31.887%   12.071/12.130/12.171/0.033        0.410/0.419/0.422/0.004       67,586,430

 avg1  34.261%                        4.906                          0.429
 tot                                662.290                         57.871      653,576,310

   text    data     bss     dec     hex filename
 136006    1312       8  137326   2186e libz-ng.so.2

```

KungFuJesus commented:
> The difference is speeds seems to be data-dependent, not cacheline-aliasing or anything like that. (I verified this using different builds to ensure things get compiled and placed slightly differently, as I often do when in doubt)

That's more or less what I'd expect. Certain data streams might hit the chunk_memcpy path harder than others, it just depends on what the decompression stream presents.  If you had a sequence of something like 17 or so bytes behind the destination pointer that repeated, I think you end up dispatching to plain old chunkcopy more in chunk_memcpy now that there are smaller unit sizes.  The inlining alone in my testing bought us like 3%, even without smaller chunks to work with.  Being able to share the last len byte copy loop shrank some of the code as well.

I have plans to make the GET_CHUNK_MAG code better for VBMI platforms as well but I'm still waiting on hardware for that.

nmoinvaz commented on chunkset_tpl.h:
```diff
@@ -115,6 +129,40 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
+
+    /* Only AVX2 */
+#ifdef HAVE_HALF_CHUNK
+    if (len <= sizeof(halfchunk_t)) {
```
I feel like a comment might be needed to sort of explain halfchunking.

KungFuJesus replied:
I added a comment about it, let me know what you think.

nmoinvaz replied:
LGTM thanks!

nmoinvaz commented on chunkset_tpl.h:
```diff
@@ -115,6 +129,40 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
+
+    /* Only AVX2 */
+#ifdef HAVE_HALF_CHUNK
+    if (len <= sizeof(halfchunk_t)) {
```
Why wouldn't we just call `chunkmemset_ssse3`, is it because it is inlined?

KungFuJesus replied:
Correct. And we're able to shrink the code size by merging common elements

KungFuJesus replied:
The inlining bought us a lot here by itself and by allowing for a half chunk length we get the best of both worlds.

KungFuJesus replied:
I added a comment about it, let me know what you think.

nmoinvaz commented:
Can this strategy also be used for AVX512?

KungFuJesus commented:
> Can this strategy also be used for AVX512?

It certainly _could_ (with a quarter size chunk being represented with `__m128i`'s).  However, on Intel ISAs, it's been repeatedly proven that the clock penalty _tends_ to hurt more than the throughput gain in extra register widths (particularly for something like a copy operation).  For Zen 5 that may not be the case, but I don't have any Zen5 hardware to test with.

With Icelake and VBMI2, that chunk magazine logic gets quite a bit simpler since they _finally_ allowed cross lane single byte permutes.  I plan to go back and add that in for 256 bit vector lengths once I can actually test it.  That might also be a good time to revisit 64 byte chunks to see if that juice is worth the squeeze.

Dead2 commented:
I did another benchmark with this PR applied to develop before #1776, added it to the post above, so it is easier to compare them.
Edit: Will have to re-do it, just noticed it was done with the AVX512 code compiled in, the others were without. It might make a difference.

KungFuJesus commented:
> I did another benchmark with this PR applied to develop before #1776, added it to the post above, so it is easier to compare them. Edit: Will have to re-do it, just noticed it was done with the AVX512 code compiled in, the others were without. It might make a difference.

Which version of gcc are you using? I feel like chunk_copy_safe has been inlining for me since at least gcc 11, but I'll try a myriad of compilation options and compilers to be sure. Earliest I can go back is probably 11 or 12.

Dead2 commented:
> > I did another benchmark with this PR applied to develop before #1776, added it to the post above, so it is easier to compare them. Edit: Will have to re-do it, just noticed it was done with the AVX512 code compiled in, the others were without. It might make a difference.
> 
> Which version of gcc are you using? I feel like chunk_copy_safe has been inlining for me since at least gcc 11, but I'll try a myriad of compilation options and compilers to be sure. Earliest I can go back is probably 11 or 12.

I don't think it has ever been inlining for me with the default CMake config. I remember disabling -Winlining a couple years ago because it was so noisy, then re-discovered it this summer when we were having a hard look at optional warnings.

I get exactly the same behavior with GCC 12.2 from Debian on Aarch64 as I do with GCC 13.2.1 from Fedora on x86-64.
I also got the same behavior with GCC 11.x before I upgraded Fedora release recently.

You don't compile using custom flags or something? -O3 for example? That always seems to slow things down for me though.

PS: Updated the benchmarks above, so now they are the correct ones. Hardly any difference from the incorrect build though.

KungFuJesus commented:
I do build with O3 and that might be the difference. Let me mess around with more stock flags and see if I see the same inlining behavior.

mtl1979 commented:
> I do build with O3 and that might be the difference. Let me mess around with more stock flags and see if I see the same inlining behavior.

As far as I know and remember, `-O3` enables some quite aggressive optimisations which for example (re-)vectorize the code and that might likely result code getting short enough to be able to be inlined. This pretty much work for anything newer than SSE3 with modern compilers as optimizer support for SSE3 and earlier iterations was dropped a long time ago.

KungFuJesus commented:
> > I do build with O3 and that might be the difference. Let me mess around with more stock flags and see if I see the same inlining behavior.
> 
> As far as I know and remember, `-O3` enables some quite aggressive optimisations which for example (re-)vectorize the code and that might likely result code getting short enough to be able to be inlined. This pretty much work for anything newer than SSE3 with modern compilers as optimizer support for SSE3 and earlier iterations was dropped a long time ago.

I don't think that's an issue as the minimum subset of extensions on all x86-64 CPUs contains sse3.

KungFuJesus commented:
> > > I did another benchmark with this PR applied to develop before #1776, added it to the post above, so it is easier to compare them. Edit: Will have to re-do it, just noticed it was done with the AVX512 code compiled in, the others were without. It might make a difference.
> > 
> > 
> > Which version of gcc are you using? I feel like chunk_copy_safe has been inlining for me since at least gcc 11, but I'll try a myriad of compilation options and compilers to be sure. Earliest I can go back is probably 11 or 12.
> 
> I don't think it has ever been inlining for me with the default CMake config. I remember disabling -Winlining a couple years ago because it was so noisy, then re-discovered it this summer when we were having a hard look at optional warnings.
> 
> I get exactly the same behavior with GCC 12.2 from Debian on Aarch64 as I do with GCC 13.2.1 from Fedora on x86-64. I also got the same behavior with GCC 11.x before I upgraded Fedora release recently.
> 
> You don't compile using custom flags or something? -O3 for example? That always seems to slow things down for me though.
> 
> PS: Updated the benchmarks above, so now they are the correct ones. Hardly any difference from the incorrect build though.

Yep, so I confirmed it's -O3 that allows this to be inlined (with that commit reverted) and in my testing, it's a significant improvement.

Dead2 commented:
> Yep, so I confirmed it's -O3 that allows this to be inlined (with that commit reverted) and in my testing, it's a significant improvement.

Problem is, -O3 is not the default for most distros, so we cannot rely on that being the case.

KungFuJesus commented:
> > Yep, so I confirmed it's -O3 that allows this to be inlined (with that commit reverted) and in my testing, it's a significant improvement.
> 
> Problem is, -O3 is not the default for most distros, so we cannot rely on that being the case.

Right but apart from manually forcing the same decisions the compiler is doing with O3 I don't know a better way to get that performance back.  Without reverting 6b8efe786, it's a net loss for me.

Dead2 commented:
> Right but apart from manually forcing the same decisions the compiler is doing with O3 I don't know a better way to get that performance back. Without reverting [6b8efe7](https://github.com/zlib-ng/zlib-ng/commit/6b8efe78685926d89da0cf1c496c1022fc24f588), it's a net loss for me.

I am open to reverting that change again, but only if we can find a good way that also improves performance on an -O2 compile.
Worst case, we could use cmake options to select between the two methods, but IMHO that is not exactly a great solution.

KungFuJesus commented:
Here's the comparison of O2 and O3, with 6b8efe7 reverted:
```
Comparing decode_real (from ./benchmark_zlib_apps_O2) to decode_real (from ./benchmark_zlib_apps_O3)
Benchmark                                                                                                Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_pvalue                 0.4735          0.4903      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_mean                  +0.0029         +0.0028          9707          9736          9712          9739
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_median                +0.0001         -0.0000          9729          9730          9733          9733
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_stddev                -0.1475         -0.1467           110            94           110            94
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_cv                    -0.1500         -0.1491             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_mean                  -0.0917         -0.0918         58686         53304         58714         53324
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_median                -0.0905         -0.0907         58636         53333         58665         53346
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_stddev                +2.2214         +2.2434           134           431           134           434
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_cv                    +2.5467         +2.5713             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_mean                  -0.0885         -0.0886         56373         51382         56399         51400
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_median                -0.0893         -0.0895         56351         51320         56377         51333
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_stddev                +1.9267         +1.9454            95           278            95           279
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_cv                    +2.2110         +2.2318             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_mean                  -0.0888         -0.0888         51371         46811         51392         46828
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_median                -0.0873         -0.0873         51332         46849         51357         46871
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_stddev                +1.5714         +1.6452           122           314           118           313
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_cv                    +1.8219         +1.9030             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_mean                  -0.0856         -0.0856         47698         43615         47714         43630
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_median                -0.0842         -0.0844         47670         43654         47690         43665
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_stddev                +1.9497         +1.9987            93           274            92           275
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_cv                    +2.2258         +2.2795             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_mean                  -0.0817         -0.0817         46265         42483         46280         42498
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_median                -0.0816         -0.0816         46218         42446         46228         42456
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_stddev                +1.6292         +1.7346           115           302           111           305
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_cv                    +1.8632         +1.9779             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_mean                  -0.0799         -0.0798         43589         40108         43603         40123
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_median                -0.0782         -0.0782         43555         40148         43576         40168
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_stddev                +0.7956         +0.8423           117           210           114           209
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_cv                    +0.9514         +1.0021             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_mean                  -0.0753         -0.0752         42907         39678         42921         39693
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_median                -0.0737         -0.0735         42873         39715         42875         39726
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_stddev                +1.2346         +1.2600            92           207            92           208
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_cv                    +1.4165         +1.4438             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_mean                  -0.0720         -0.0720         41604         38611         41620         38623
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_median                -0.0717         -0.0718         41594         38609         41608         38619
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_stddev                +1.0171         +1.0073            90           181            90           180
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_cv                    +1.1735         +1.1631             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_mean                  -0.0772         -0.0772         41126         37951         41140         37964
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_median                -0.0760         -0.0762         41058         37936         41077         37945
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_stddev                +0.6152         +0.6330           125           202           124           203
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_cv                    +0.7503         +0.7696             0             0             0             0
OVERALL_GEOMEAN                                                                                       -0.0741         -0.0742             0             0             0             0
```

Here's O2 with and without the revert compared:
```
Comparing decode_real (from ./benchmark_zlib_apps_O2) to decode_real (from ./benchmark_zlib_apps_splitchunkcopy)
Benchmark                                                                                                Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_pvalue                 0.0337          0.0337      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_mean                  +0.0046         +0.0046          9717          9761          9721          9765
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_median                +0.0078         +0.0078          9685          9760          9687          9762
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_stddev                -0.1403         -0.1412            76            65            76            65
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/0_cv                    -0.1442         -0.1451             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_pvalue                 0.0084          0.0071      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_mean                  -0.0019         -0.0018         58619         58509         58639         58531
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_median                -0.0010         -0.0012         58579         58520         58604         58536
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_stddev                -0.1814         -0.1930           124           101           124           100
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/1_cv                    -0.1799         -0.1915             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_mean                  -0.0029         -0.0029         56388         56223         56410         56244
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_median                -0.0027         -0.0025         56368         56215         56378         56235
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_stddev                -0.3585         -0.3613           111            71           115            73
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/2_cv                    -0.3566         -0.3594             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_mean                  +0.0029         +0.0029         51431         51579         51450         51598
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_median                +0.0046         +0.0042         51341         51578         51363         51582
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_stddev                -0.7115         -0.7044           309            89           305            90
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/3_cv                    -0.7123         -0.7053             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_mean                  -0.0054         -0.0055         47771         47511         47791         47527
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_median                -0.0048         -0.0049         47741         47510         47763         47529
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_stddev                -0.5606         -0.5436           116            51           116            53
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/4_cv                    -0.5582         -0.5410             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_mean                  -0.0107         -0.0107         46309         45812         46326         45829
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_median                -0.0108         -0.0107         46296         45798         46304         45810
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_stddev                +0.0024         +0.0119            72            72            70            71
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/5_cv                    +0.0133         +0.0229             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_mean                  -0.0177         -0.0177         43591         42820         43607         42837
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_median                -0.0177         -0.0177         43559         42789         43578         42809
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_stddev                +0.1162         +0.0773            87            97            88            95
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/6_cv                    +0.1363         +0.0967             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_mean                  -0.0191         -0.0191         42958         42139         42973         42153
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_median                -0.0191         -0.0188         42929         42109         42931         42124
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_stddev                +0.0696         +0.0752            96           103            96           103
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/7_cv                    +0.0904         +0.0961             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_mean                  -0.0198         -0.0198         41611         40788         41626         40802
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_median                -0.0184         -0.0187         41565         40798         41585         40807
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_stddev                +0.0662         +0.0655           104           111           103           110
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/8_cv                    +0.0877         +0.0871             0             0             0             0
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_mean                  -0.0196         -0.0196         41083         40278         41097         40293
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_median                -0.0183         -0.0185         41035         40285         41054         40295
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_stddev                -0.0955         -0.0654           111           101           109           102
png_[decode_real vs. decode_real]istic/png_[decode_real vs. decode_real]istic/9_cv                    -0.0774         -0.0467             0             0             0             0
OVERALL_GEOMEAN                                                                                       -0.0090         -0.0090             0             0             0             0
```

KungFuJesus commented:
Couldn't we just alter our cmake configuration to not override O3 with O2 for release?  I don't know how many distros override these things to begin with but I imagine most don't try to override the default build arguments.  For me, the difference between O2 and O3 is ~7% and the work to split chunk copy so that it partially inlines hurts the O3 performance by another 2ish%.

nmoinvaz commented:
I would be in favor of forcing CMake/configure to use -O3 or /O3 for this project.

ptr1337 commented:
> I would be in favor of forcing CMake/configure to use -O3 or /O3 for this project.

Archlinux is actually tending to use None as buildtype, so maybe the maintainer should be then contacted there.
See here:
https://gitlab.archlinux.org/archlinux/packaging/packages/zlib-ng/-/blob/main/PKGBUILD?ref_type=heads#L31

Dead2 commented:
@ptr1337 I see they also still disable UNALIGNED_OK, even though that is not needed anymore (Fixed in 2.0.7, and improved upon for 2.1.0). That probably reduces performance by a lot, since that parameter is used by many of the (safe) compiler intrinsics based optimizations.

Not sure what setting buildtype to none does, but I assume it in practice means don't use any of the cmake-provided default cflags ('-O(2|3) -NDEBUG') and just use the ones that the distro provides instead.

For the record, I have been testing -O2 vs -O3 a lot the past weeks, and thus far -O3 always hurts performance in my tests. So it seems that this might be cpu (i9-9900K, rpi3, rpi5) and at least partially data dependent.
One theory I have is that possibly the cpu-mitigations for my i9-9900K cpu hurts performance due to mitigations doing cache flushing, and -O3 uses more (instruction)cache. @KungFuJesus has a newer cpu where it performs better, and it does not have as many cpu mitigations enabled, but he also has twice as much L1 cache, so that also can help with the bigger code.

In my tests, the following gives the best performance: "-O2 -fgcse-after-reload -ftree-partial-pre -fno-tree-tail-merge". The two first options after O2 are from O3, the last one disables an optimization enabled in O1 and higher. (no-tree-tail-merge tries to deduplicate the end of functions by jumping to another that is identical, this might be slower due to jumping to somewhere not in the cache. Strangely by disabling this option, the library size goes down instead of up, so it might possibly be a bug in GCC 13.2 or just a peculiarity with our code.) Using this set of options should also be completely safe. In my tests on the i9-9900K they make inflate 1.00% faster and deflate 0.37% faster.

We are working on improving performance for both my tests and the ones @KungFuJesus does though, hopefully providing a good middle ground that helps everyone. Anyone wanting to contribute can find us via the contributing page on the wiki.

nmoinvaz commented on chunkset_tpl.h:
```diff
@@ -130,36 +179,41 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
+
+    while (len >= (2 * sizeof(chunk_t))) {
+        storechunk(out, &chunk_load);
+        storechunk(out + adv_amount, &chunk_load);
```
Was `out + sizeof(chunk_t)` a bug previously?

nmoinvaz commented on arch/x86/chunkset_avx2.c:
```diff
@@ -68,6 +74,10 @@ static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
     *chunk = _mm256_set1_epi64x(tmp);
 }
 
+static inline void chunkmemset_16(uint8_t *from, chunk_t *chunk) {
```
So we were missing one?

KungFuJesus replied:
Sort of.  It had been handled in the GET_CHUNK_MAG function but it was better to handle it here.

nmoinvaz reviewed (commented):
Code looks clean to me.

KungFuJesus commented on chunkset_tpl.h:
```diff
@@ -115,8 +125,47 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
+            out += adv_amount;
+        }
+
+        chunk_load = halfchunk2whole(halfchunk_load);
+        goto rem_bytes;
```
Doing so makes it harder for the compiler to inline this function.

nmoinvaz commented:
## Benchmark
Dataset: silesia.tar

### Develop 8d10c309
```
OS: Windows 10 10.0.22631 AMD64
CPU: Intel64 Family 6 Model 140 Stepping 1, GenuineIntel
Tool: ../zlib-ng/build-develop/Release/minigzip.exe Size: 105,472 B
Levels: 1-9
Runs: 70         Trim worst: 30

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      0.952/0.983/1.000/0.010        0.451/0.470/0.488/0.011       94,127,497
 2     35.519%      1.472/1.507/1.529/0.014        0.446/0.460/0.487/0.010       75,286,322
 3     33.844%      1.993/2.041/2.064/0.018        0.420/0.437/0.447/0.007       71,735,218
 4     33.146%      2.257/2.323/2.355/0.019        0.411/0.430/0.444/0.008       70,255,223
 5     32.642%      2.500/2.553/2.581/0.019        0.414/0.428/0.456/0.010       69,187,419
 6     32.483%      2.948/3.015/3.048/0.023        0.405/0.423/0.441/0.009       68,850,776
 7     32.255%      3.920/3.996/4.039/0.028        0.408/0.423/0.435/0.006       68,366,759
 8     32.167%      6.149/6.315/6.386/0.041        0.414/0.426/0.442/0.008       68,180,762
 9     31.887%      6.790/6.951/7.012/0.046        0.409/0.418/0.432/0.007       67,586,442

 avg1  34.261%                        3.298                          0.435
 tot                               1187.367                        156.614      653,576,418
```

### PR efb588d2
```
OS: Windows 10 10.0.22631 AMD64
CPU: Intel64 Family 6 Model 140 Stepping 1, GenuineIntel
Tool: ../zlib-ng/build-develop-jesus/Release/minigzip.exe Size: 105,984 B
Levels: 1-9
Runs: 70         Trim worst: 30

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      0.964/0.980/0.996/0.009        0.450/0.460/0.472/0.006       94,127,497
 2     35.519%      1.488/1.502/1.520/0.010        0.444/0.452/0.463/0.005       75,286,322
 3     33.844%      2.015/2.035/2.053/0.012        0.425/0.434/0.441/0.005       71,735,218
 4     33.146%      2.301/2.327/2.346/0.012        0.418/0.426/0.433/0.004       70,255,223
 5     32.642%      2.527/2.557/2.578/0.012        0.415/0.423/0.432/0.006       69,187,419
 6     32.483%      2.992/3.017/3.034/0.012        0.411/0.419/0.426/0.005       68,850,776
 7     32.255%      3.966/4.004/4.031/0.015        0.414/0.422/0.431/0.005       68,366,759
 8     32.167%      6.274/6.310/6.350/0.020        0.414/0.419/0.425/0.003       68,180,762
 9     31.887%      6.858/6.920/6.951/0.020        0.406/0.415/0.422/0.005       67,586,442

 avg1  34.261%                        3.295                          0.430
 tot                               1186.072                        154.739      653,576,418
```

KungFuJesus commented on arch/x86/chunkset_avx2.c:
```diff
@@ -118,6 +125,51 @@ static inline chunk_t GET_CHUNK_MAG(uint8_t *buf, uint32_t *chunk_rem, uint32_t
+    return _mm256_zextsi128_si256(chunk);
+}
+
+static inline halfchunk_t GET_HALFCHUNK_MAG(uint8_t *buf, uint32_t *chunk_rem, uint32_t dist) {
+    lut_rem_pair lut_rem = perm_idx_lut[dist - 3];
+    __m128i perm_vec, ret_vec;
+    __msan_unpoison(buf + dist, 16 - dist);
+    ret_vec = _mm_loadu_si128((__m128i*)buf);
+    *chunk_rem = half_rem_vals[dist - 3];
+
+    perm_vec = _mm_load_si128((__m128i*)(permute_table + lut_rem.idx));
+    ret_vec = _mm_shuffle_epi8(ret_vec, perm_vec);
+
+    return ret_vec;
+}
```
The branching cost far outweighs the savings from the load.

Dead2 reviewed (approved):
LGTM
